backend/copr_backend/background_worker_build.py

```diff
@@ -740,6 +766,122 @@ def _add_pubkey(self):
         self.log.info("Added pubkey for user %s project %s into: %s",
                       user, project, pubkey_path)
 
+    @skipped_for_source_build
+    def _setup_for_user_ssh(self):
+        """
+        Setup the builder for user SSH
+        https://github.com/fedora-copr/debate/tree/main/user-ssh-builders
+
+        If the builder setup for user SSH becomes more complicated than just
+        installing the public key, we might want to move the code to a script
+        within `copr-builder` and call it here or from `copr-rpmbuild`. There
+        is no requirement for it to be here.
+        """
+        if not self.job.ssh_public_keys:
+            return
+        self._alloc_root_ssh_connection()
+        self._deploy_user_ssh()
+        self._set_default_expiration()
+
+    def _alloc_root_ssh_connection(self):
+        self.log.info("Allocating root ssh connection to builder")
+        self.root_ssh = SSHConnection(
+            user="root",
+            host=self.host.hostname,
+            config_file=self.opts.ssh.builder_config,
+            log=self.log,
+        )
+
+    def _deploy_user_ssh(self):
+        """
+        Deploy user public key to the builder, so that they can connect via SSH.
+        """
+        pubkey = shlex.quote(self.job.ssh_public_keys)
+        cmd = COMMANDS["echo_authorized_keys"].format(pubkey)
+        rc, _out, _err = self.root_ssh.run_expensive(cmd)
+        if rc != 0:
+            self.log.error("Failed to deploy user SSH key for %s",
+                           self.job.project_owner)
+            return
+        self.log.info("Deployed user SSH key for %s", self.job.project_owner)
+
+    def _set_default_expiration(self):
+        """
+        Set the default expiration time for the builder
+        """
+        default = self.job.started_on + USER_SSH_DEFAULT_EXPIRATION
+        cmd = COMMANDS["set_expiration"].format(shlex.quote(str(default)))
+        rc, _out, _err = self.root_ssh.run_expensive(cmd)
+        if rc != 0:
+            # This only affects the `copr-builder show` command to print unknown
+            # remaining time. It won't affect the backend in terminating the
+            # buidler when it is supposed to
+            self.log.error("Failed to set the default expiration time")
+            return
+        self.log.info("The expiration time was set to %s", default)
+
+    def _builder_expiration(self):
+        """
+        Find the user preference for the builder expiration.
+        """
+        rc, out, _err = self.root_ssh.run_expensive(COMMANDS["cat_expiration"])
```

Comment:
Don't use `run_expensive` and check output length and/or timeout.

Reply:
I couldn't use `run` because it doesn't return the output. So I kept `run_expensive` and added a timeout for it.